### PR TITLE
Fix memory leak in Vector move assignment

### LINKED ISSUE #4499: Memory leak with Memory move assignment operator?
I've tracked a memory leak in application code down to the move assignment operator in the `Memory` class.

I effectively have a function like
```c++
Vector doStuff(const Vector &input)
{
  Vector output;
  // do stuff
  return output;
}
```

If my function is called with an assignment to an uninitialized vector like
```c++
Vector a({1,2,3,4});
Vector b = doStuff(a);
```
things are all great.  If my function is called to an initialized vector like
```c++
Vector a({1,2,3,4});
Vector b(5);
b = doStuff(a);
```
the 5 original doubles for `b` become a memory leak because the corresponding move assignment operator for `Vector` [is](https://github.com/mfem/mfem/blob/957ca0d50089795b219e4701312355f875af6019/linalg/vector.cpp#L143)
```c++
Vector &Vector::operator=(Vector &&v)
{
   data = std::move(v.data);
   // Self-assignment-safe way to move v.size to size:
   const auto size_tmp = v.size;
   v.size = 0;
   size = size_tmp;
   return *this;
}
```
and the move assignment operator for `Memory` [is](https://github.com/mfem/mfem/blob/957ca0d50089795b219e4701312355f875af6019/general/mem_manager.hpp#L227)
```c++
Memory &operator=(Memory &&orig)
{
   // Guard self-assignment:
   if (this == &orig) { return *this; }
   *this = orig;
   orig.Reset();
   return *this;
}
```
  
While I could modify my function signature to `void doStuff(const Vector &input, Vector &output)`, I can see others wanting to use something like the original function signature to leverage the move assignment operator for `Vector`.  My naive fix for this is to add a line to the move assignment operator for `Memory` to delete the old data, i.e.,
```c++
Memory &operator=(Memory &&orig)
{
   // Guard self-assignment:
   if (this == &orig) { return *this; }
   Delete();
   *this = orig;
   orig.Reset();
   return *this;
}
```
but I can imagine this may cause some issues elsewhere?

### PR DESCRIPTION
**Pull request by pazner**
Resolves #4499.

Two notes:

* The check for `if (this != &v)` is not technically required, since self-move leaves the object in a valid but unspecified state ([see here](https://en.cppreference.com/w/cpp/utility/move#Notes)).
* We also don't strictly need to destroy the moved-from object, but it's probably clearer to do so.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4500](https://github.com/mfem/mfem/pull/4500) | @pazner | @tzanio | @paul-hilscher + @cjvogl | 9/15/24 | 9/21/24 | 9/25/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Review by cjvogl — APPROVED** [review]
Looks good to me (i.e., addresses the memory leak), thanks @pazner!

```
==2549== All heap blocks were freed -- no leaks are possible
```

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by paul-hilscher — APPROVED** [review]
nice fix! Thanks.

File: linalg/vector.cpp [hunk 1]
```diff
@@ -142,11 +142,8 @@ Vector &Vector::operator=(const Vector &v)
 
 Vector &Vector::operator=(Vector &&v)
 {
-   data = std::move(v.data);
-   // Self-assignment-safe way to move v.size to size:
-   const auto size_tmp = v.size;
-   v.size = 0;
-   size = size_tmp;
+   v.Swap(*this);
```
Comment (paul-hilscher) [inline linalg/vector.cpp:145]: small question (outside of this PR):

The swap operation in `vector.hpp` is given by
```
inline void Vector::Swap(Vector &other)
{
   mfem::Swap(data, other.data);
   mfem::Swap(size, other.size);
}
```
The implementation of `Swap` is found in `array.hpp`:
```
template <class T>
inline void Swap(T &a, T &b)
{
   T c = a;
   a = b;
   b = c;
}
```

1. My current understanding is that this implementation would use copy assignment for `Memory`:
```
   /// Copy-assignment operator: default.
   Memory &operator=(const Memory &orig) = default;

   /** Move assignment operator. Sets the pointers and associated ownership of
       validity flags of @a *this to those of @a other. Resets @a other. */
   Memory &operator=(Memory &&orig)
```
2. Would it make sense to migrate migrate to [std::swap](https://en.cppreference.com/w/cpp/algorithm/swap) which supports move assignments ? Happy to look into this further if there is interest.

Comment (pazner) [reply linalg/vector.cpp:145]: Copying a memory object is cheap, it's similar to copying a pointer (there is no underlying memcpy happening).

(But yes, I think we should either use `std::move` in `mfem::Swap` or just call `std::swap`. I actually made that change in PR #4028).

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**